pkg/experiment/metastore/index/index.go

```diff
@@ -0,0 +1,511 @@
+func (i *Index) CreatePartitionKey(blockId string) PartitionKey {
+	t := ulid.Time(ulid.MustParse(blockId).Time()).UTC()
+
+	var b strings.Builder
+	b.Grow(16)
+
+	year, month, day := t.Date()
+	b.WriteString(fmt.Sprintf("%04d%02d%02d", year, month, day))
+
+	partitionDuration := i.Config.PartitionDuration
+	if partitionDuration < 24*time.Hour {
+		hour := (t.Hour() / int(partitionDuration.Hours())) * int(partitionDuration.Hours())
+		b.WriteString(fmt.Sprintf("T%02d", hour))
+	}
+
+	mDuration := model.Duration(partitionDuration)
+	b.WriteString(".")
+	b.WriteString(mDuration.String())
+
+	return PartitionKey(b.String())
+}
+
+// FindPartitionMeta retrieves the partition meta for the given key.
+func (i *Index) FindPartitionMeta(key PartitionKey) *PartitionMeta {
+	loaded, ok := i.loadedPartitions[key]
+	if ok {
+		return loaded.meta
+	}
+	for _, p := range i.allPartitions {
+		if p.Key == key {
+			return p
+		}
+	}
+	return nil
+}
+
+// InsertBlock is the primary way for adding blocks to the index.
+func (i *Index) InsertBlock(b *metastorev1.BlockMeta) error {
+	i.partitionMu.Lock()
+	defer i.partitionMu.Unlock()
+
+	return i.insertBlock(b)
+}
+
+// insertBlock is the underlying implementation for inserting blocks. It is the caller's responsibility to enforce safe
+// concurrent access. The method will create a new partition if needed.
+func (i *Index) insertBlock(b *metastorev1.BlockMeta) error {
+	meta, err := i.GetOrCreatePartitionMeta(b)
+	if err != nil {
+		return err
+	}
+
+	p := i.getOrCreatePartition(meta)
+
+	p.shardsMu.Lock()
+	defer p.shardsMu.Unlock()
+
+	s, ok := p.shards[b.Shard]
+	if !ok {
+		s = &indexShard{
+			tenants: make(map[string]*indexTenant),
+		}
+		p.shards[b.Shard] = s
+	}
+
+	s.tenantsMu.Lock()
+	defer s.tenantsMu.Unlock()
+
+	ten, ok := s.tenants[b.TenantId]
+	if !ok {
+		ten = &indexTenant{
+			blocks: make(map[string]*metastorev1.BlockMeta),
+		}
+		s.tenants[b.TenantId] = ten
+	}
+
+	ten.blocksMu.Lock()
+	defer ten.blocksMu.Unlock()
+
+	ten.blocks[b.Id] = b
+	return nil
+}
+
+// GetOrCreatePartitionMeta makes the mapping between blocks and partitions. It may assign the block to an existing
+// partition or create a new partition altogether. Meant to be used only in the context of new blocks.
+func (i *Index) GetOrCreatePartitionMeta(b *metastorev1.BlockMeta) (*PartitionMeta, error) {
+	key := i.CreatePartitionKey(b.Id)
+	meta := i.FindPartitionMeta(key)
+
+	if meta == nil {
+		ts, duration, err := key.Parse()
+		if err != nil {
+			return nil, err
+		}
+		meta = &PartitionMeta{
+			Key:       key,
+			Ts:        ts,
+			Duration:  duration,
+			Tenants:   make([]string, 0),
+			tenantMap: make(map[string]struct{}),
+		}
+		i.allPartitions = append(i.allPartitions, meta)
+		i.sortPartitions()
+	}
+
+	if b.TenantId != "" {
+		meta.AddTenant(b.TenantId)
+	} else {
+		for _, ds := range b.Datasets {
+			meta.AddTenant(ds.TenantId)
+		}
+	}
+
+	return meta, nil
+}
```

Comment:
It looks like we need to take the global index mutex here. Despite the fact that `GetOrCreatePartitionMeta` is only called in the context of a raft command, i.e. synchronously and without a risk that someone else will add the partition right after `FindPartitionMeta` returning a nil, it may race with access to `allPartitions` from the read path functions.